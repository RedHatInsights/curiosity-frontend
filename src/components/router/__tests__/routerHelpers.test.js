import {
  baseName,
  dynamicBaseName,
  getErrorRoute,
  getNavigationDetail,
  getRouteDetail,
  getNavRouteDetail
} from '../routerHelpers';

describe('RouterHelpers', () => {
  const mockWindowLocationProp = async ({ prop, url, callback }) => {
    const { location } = window;
    delete window.location;
    window.location = { [prop]: url };
    await callback();
    window.location = location;
  };

  it('should return specific properties', () => {
    expect(baseName).toBeDefined();
    expect(dynamicBaseName).toBeDefined();
    expect(getErrorRoute).toBeDefined();
    expect(getNavigationDetail).toBeDefined();
    expect(getRouteDetail).toBeDefined();
    expect(getNavRouteDetail).toBeDefined();
  });

  it('should return a generated baseName using NO path prefix', () => {
    expect(
      dynamicBaseName({
        pathName: '/appName',
        pathPrefix: ''
      })
    ).toMatchSnapshot('empty pathPrefix: app base name');

    expect(
      dynamicBaseName({
        pathName: '/appName'
      })
    ).toMatchSnapshot('undefined pathPrefix: app base name');

    expect(
      dynamicBaseName({
        pathName: '/appName',
        pathPrefix: '/'
      })
    ).toMatchSnapshot('root reference pathPrefix: app base name');

    expect(
      dynamicBaseName({
        pathName: '/appName/loremRoute',
        pathPrefix: ''
      })
    ).toMatchSnapshot('app lorem route name');
  });

  it('should return a generated baseName using a beta path prefix', () => {
    expect(
      dynamicBaseName({
        pathName: '/appName',
        pathPrefix: '/beta'
      })
    ).toMatchSnapshot('app base name');

    expect(
      dynamicBaseName({
        pathName: '/beta/appName',
        pathPrefix: '/beta'
      })
    ).toMatchSnapshot('beta app base name');

    expect(
      dynamicBaseName({
        pathName: '/beta/appName/loremRoute',
        pathPrefix: '/beta'
      })
    ).toMatchSnapshot('beta app lorem route name');
  });

  it('should return an error route', () => {
    expect(getErrorRoute).toMatchSnapshot('error route');
  });

  it('should return navigation and route details that align to location', () => {
    expect({
      nav: getNavigationDetail({ id: 'soon' }),
      route: getRouteDetail({ id: 'soon' }),
      navRoute: getNavRouteDetail({ id: 'soon' })
    }).toMatchSnapshot('detail: specific route ID');

    expect({
      nav: getNavigationDetail({ id: 'arm' }),
      route: getRouteDetail({ id: 'arm' }),
      navRoute: getNavRouteDetail({ id: 'arm' })
    }).toMatchSnapshot('detail: specific navigation ID');

    expect({
      nav: getNavigationDetail({ pathname: '/rhel-sw/all' }),
      route: getRouteDetail({ pathname: '/rhel-sw/all' }),
      navRoute: getNavRouteDetail({ pathname: '/rhel-sw/all' })
    }).toMatchSnapshot('detail: match specific path navigation');

    expect({
      nav: getNavigationDetail({ id: 'lorem-missing', pathname: '/rhel-sw/all' }),
      route: getRouteDetail({ id: 'lorem-missing', pathname: '/rhel-sw/all' }),
      navRoute: getNavRouteDetail({ id: 'lorem-missing', pathname: '/rhel-sw/all' })
    }).toMatchSnapshot('detail: missing ID, specific path');

    expect({
      nav: getNavigationDetail({ id: 'lorem', pathname: '/lorem-ipsum-missing', returnDefault: true }),
      route: getRouteDetail({ id: 'lorem', pathname: '/lorem-ipsum-missing', returnDefault: true }),
      navRoute: getNavRouteDetail({ id: 'lorem', pathname: '/lorem-ipsum-missing', returnDefault: true })
    }).toMatchSnapshot('detail: missing id and pathname, default');

    expect({
      nav: getNavigationDetail({ id: 'lorem', pathname: '/lorem-ipsum-missing', returnDefault: false }),
      route: getRouteDetail({ id: 'lorem', pathname: '/lorem-ipsum-missing', returnDefault: false }),
      navRoute: getNavRouteDetail({ id: 'lorem', pathname: '/lorem-ipsum-missing', returnDefault: false })
    }).toMatchSnapshot('detail: missing id and pathname and default');

    expect({
      nav: getNavigationDetail({}),
      route: getRouteDetail({}),
      navRoute: getNavRouteDetail({})
    }).toMatchSnapshot('detail: missing parameters');
  });

  it('should handle location search and hash passthrough values', () => {
    mockWindowLocationProp({
      prop: 'href',
      url: 'https://ci.foo.redhat.com/subscriptions/rhel-sw/all',
      callback: () => {
        expect({
          routeHref: getNavigationDetail({ pathname: '/rhel-sw/all' }).routeHref
        }).toMatchSnapshot('NO search and hash');
      }
    });

    mockWindowLocationProp({
      prop: 'href',
      url: 'https://ci.foo.redhat.com/subscriptions/rhel-sw/all?dolor=sit',
      callback: () => {
        expect({
          routeHref: getNavigationDetail({ pathname: '/rhel-sw/all' }).routeHref
        }).toMatchSnapshot('search');
      }
    });

    mockWindowLocationProp({
      prop: 'href',
      url: 'https://ci.foo.redhat.com/subscriptions/rhel-sw/all#lorem',
      callback: () => {
        expect({
          routeHref: getNavigationDetail({ pathname: '/rhel-sw/all' }).routeHref
        }).toMatchSnapshot('hash');
      }
    });

    mockWindowLocationProp({
      prop: 'href',
      url: 'https://ci.foo.redhat.com/subscriptions/rhel-sw/all?dolor=sit#lorem',
      callback: () => {
        expect({
          routeHref: getNavigationDetail({ pathname: '/rhel-sw/all' }).routeHref
        }).toMatchSnapshot('search and hash');
      }
    });
  });
});
