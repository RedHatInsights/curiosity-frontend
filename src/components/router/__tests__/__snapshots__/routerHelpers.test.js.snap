// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterHelpers should handle location search and hash passthrough values: NO search and hash 1`] = `
Object {
  "routeHref": "/rhel-sw/all",
}
`;

exports[`RouterHelpers should handle location search and hash passthrough values: hash 1`] = `
Object {
  "routeHref": "/rhel-sw/all#lorem",
}
`;

exports[`RouterHelpers should handle location search and hash passthrough values: search 1`] = `
Object {
  "routeHref": "/rhel-sw/all?dolor=sit",
}
`;

exports[`RouterHelpers should handle location search and hash passthrough values: search and hash 1`] = `
Object {
  "routeHref": "/rhel-sw/all?dolor=sit#lorem",
}
`;

exports[`RouterHelpers should return a generated baseName using NO path prefix: app lorem route name 1`] = `"/appName"`;

exports[`RouterHelpers should return a generated baseName using NO path prefix: empty pathPrefix: app base name 1`] = `"/appName"`;

exports[`RouterHelpers should return a generated baseName using NO path prefix: root reference pathPrefix: app base name 1`] = `"/appName"`;

exports[`RouterHelpers should return a generated baseName using NO path prefix: undefined pathPrefix: app base name 1`] = `"/appName"`;

exports[`RouterHelpers should return a generated baseName using a beta path prefix: app base name 1`] = `"/appName"`;

exports[`RouterHelpers should return a generated baseName using a beta path prefix: beta app base name 1`] = `"/beta/appName"`;

exports[`RouterHelpers should return a generated baseName using a beta path prefix: beta app lorem route name 1`] = `"/beta/appName"`;

exports[`RouterHelpers should return an error route: error route 1`] = `
Object {
  "activateOnError": true,
  "component": Object {
    "$$typeof": Symbol(react.memo),
    "WrappedComponent": [Function],
    "compare": null,
    "displayName": "Connect(OptinView)",
    "type": [Function],
  },
  "disabled": false,
  "exact": true,
  "id": "optin",
  "render": true,
  "to": "/optin",
}
`;

exports[`RouterHelpers should return navigation and route details that align to location: detail: match specific path navigation 1`] = `
Object {
  "nav": Object {
    "default": true,
    "id": "all",
    "path": "/rhel-sw/all",
    "pathId": "rhel-sw",
    "pathParameter": "RHEL",
    "routeHref": "/rhel-sw/all",
  },
  "navRoute": Object {
    "component": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "displayName": "Connect(RhelView)",
      "type": [Function],
    },
    "default": true,
    "disabled": false,
    "exact": true,
    "id": "all",
    "path": "/rhel-sw/all",
    "pathId": "rhel-sw",
    "pathParameter": "RHEL",
    "redirect": true,
    "render": true,
    "routeHref": "/rhel-sw/all",
    "to": "/rhel-sw/all",
  },
  "route": Object {
    "component": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "displayName": "Connect(RhelView)",
      "type": [Function],
    },
    "disabled": false,
    "exact": true,
    "redirect": true,
    "render": true,
    "to": "/rhel-sw/all",
  },
}
`;

exports[`RouterHelpers should return navigation and route details that align to location: detail: missing ID, specific path 1`] = `
Object {
  "nav": Object {
    "default": true,
    "id": "all",
    "path": "/rhel-sw/all",
    "pathId": "rhel-sw",
    "pathParameter": "RHEL",
    "routeHref": "/rhel-sw/all",
  },
  "navRoute": Object {
    "component": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "displayName": "Connect(RhelView)",
      "type": [Function],
    },
    "default": true,
    "disabled": false,
    "exact": true,
    "id": "all",
    "path": "/rhel-sw/all",
    "pathId": "rhel-sw",
    "pathParameter": "RHEL",
    "redirect": true,
    "render": true,
    "routeHref": "/rhel-sw/all",
    "to": "/rhel-sw/all",
  },
  "route": Object {
    "component": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "displayName": "Connect(RhelView)",
      "type": [Function],
    },
    "disabled": false,
    "exact": true,
    "redirect": true,
    "render": true,
    "to": "/rhel-sw/all",
  },
}
`;

exports[`RouterHelpers should return navigation and route details that align to location: detail: missing id and pathname and default 1`] = `
Object {
  "nav": Object {},
  "navRoute": Object {},
  "route": Object {},
}
`;

exports[`RouterHelpers should return navigation and route details that align to location: detail: missing id and pathname, default 1`] = `
Object {
  "nav": Object {
    "default": true,
    "id": "all",
    "path": "/rhel-sw/all",
    "pathId": "rhel-sw",
    "pathParameter": "RHEL",
    "routeHref": "/rhel-sw/all",
  },
  "navRoute": Object {
    "component": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "displayName": "Connect(RhelView)",
      "type": [Function],
    },
    "default": true,
    "disabled": false,
    "exact": true,
    "id": "all",
    "path": "/rhel-sw/all",
    "pathId": "rhel-sw",
    "pathParameter": "RHEL",
    "redirect": true,
    "render": true,
    "routeHref": "/rhel-sw/all",
    "to": "/rhel-sw/all",
  },
  "route": Object {},
}
`;

exports[`RouterHelpers should return navigation and route details that align to location: detail: missing parameters 1`] = `
Object {
  "nav": Object {},
  "navRoute": Object {},
  "route": Object {},
}
`;

exports[`RouterHelpers should return navigation and route details that align to location: detail: specific navigation ID 1`] = `
Object {
  "nav": Object {
    "id": "arm",
    "path": "/rhel-sw/arm",
    "pathId": "rhel-sw",
    "pathParameter": "RHEL for ARM",
    "routeHref": "/rhel-sw/arm",
  },
  "navRoute": Object {
    "component": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "displayName": "Connect(RhelView)",
      "type": [Function],
    },
    "disabled": false,
    "exact": true,
    "id": "arm",
    "path": "/rhel-sw/arm",
    "pathId": "rhel-sw",
    "pathParameter": "RHEL for ARM",
    "render": true,
    "routeHref": "/rhel-sw/arm",
    "to": "/rhel-sw/:variant",
  },
  "route": Object {},
}
`;

exports[`RouterHelpers should return navigation and route details that align to location: detail: specific route ID 1`] = `
Object {
  "nav": Object {},
  "navRoute": Object {},
  "route": Object {
    "component": Object {
      "$$typeof": Symbol(react.memo),
      "WrappedComponent": [Function],
      "compare": null,
      "displayName": "Connect(TourView)",
      "type": [Function],
    },
    "disabled": false,
    "exact": true,
    "id": "soon",
    "render": true,
    "to": "/soon",
  },
}
`;
